### ISSUE TITLE
false positive `reportUnreachable` (instances can be falsey)

### ISSUE THREAD

**Issue by KotlinIsland** [DetachHead/basedpyright#666]
```py
class A: ...
class B(A):
    def __bool__(self):
        return False
def f(a: A):
    if not a:
        print("hi")  # error: code is unreachable

f(B())
```
[playground](https://basedpyright.com/?typeCheckingMode=all&code=MYGwhgzhAECCBc0B0KBQpIwEIApYEp5VoToATAUwDNoB9WgIwHsmR6cIKQrDjT%2BAThQAuAVwEA7aADEwITqko0qOMIgJF%2BASxoSmw6Gr78SABwFaJwnACIAFlpv5UqFbnzOgA)

although, this should still be reported here, but with a different code/message: "A does not define `__bool__` or `__len__` so will always be `True` here, but a subtype could implement that and become `False`"

**Comment by leiserfg** [comment]
Something similar happens with django-rest-framework, it has a `get_serializer_class` method that contains a 
`assert self.serializer_class is not None` 'cause one is supposed to set that field in the class (or in a parent of it). Then pyright (and based) complaints, even if such field is defined.

**Comment by DetachHead** [comment]
that sounds like a different issue because `is None`/`is not None` checks are not effected by the `__bool__` method. would you mind poviding a minimal example either here or in a new issue? thanks

**Comment by DetachHead** [comment]
oh wow, that should be an error on the definition of `B.value`, because it's incompatible with the supertype.

i've raised a separate issue for this: #808
